tests/trx_generator/trx_generator.cpp

```diff
@@ -202,4 +205,117 @@ namespace eosio::testing {
       return true;
    }
 
-}
\ No newline at end of file
+   fc::variant json_from_file_or_string(const string& file_or_str, fc::json::parse_type ptype = fc::json::parse_type::legacy_parser)
+   {
+      regex r("^[ \t]*[\{\[]");
+      if ( !regex_search(file_or_str, r) && fc::is_regular_file(file_or_str) ) {
+         try {
+            return fc::json::from_file(file_or_str, ptype);
+         } EOS_RETHROW_EXCEPTIONS(json_parse_exception, "Fail to parse JSON from file: ${file}", ("file", file_or_str));
+
+      } else {
+         try {
+            return fc::json::from_string(file_or_str, ptype);
+         } EOS_RETHROW_EXCEPTIONS(json_parse_exception, "Fail to parse JSON from string: ${string}", ("string", file_or_str));
+      }
+   }
+
+   trx_generator::trx_generator(std::string chain_id_in, const std::string& abi_data_file, std::string contract_owner_account, std::string auth_account, std::string action_name,
+         const std::string& action_data_file_or_str, int64_t trx_expr, const std::string& private_key_str, std::string lib_id_str, std::string log_dir) :
+      _provider(), _chain_id(chain_id_in), _abi_data_file_path(abi_data_file), _contract_owner_account(contract_owner_account), _auth_account(auth_account), _action(action_name), _action_data_file_or_str(action_data_file_or_str),
+      _trx_expiration(trx_expr*1000000), _private_key(fc::crypto::private_key(private_key_str)),
+      _last_irr_block_id(fc::variant(lib_id_str).as<block_id_type>()), _log_dir(log_dir)
+      {
+      }
+
+   bool trx_generator::setup() {
+      _nonce_prefix = 0;
+      _nonce = static_cast<uint64_t>(fc::time_point::now().sec_since_epoch()) << 32;
+
+      std::cout
+            << "Stop Generation (form potential ongoing generation in preparation for starting new generation run)."
+            << std::endl;
+      stop_generation();
+
+      std::cout << "Create Initial Transaction with action data." << std::endl;
+      abi_serializer abi = abi_serializer(fc::json::from_file(_abi_data_file_path).as<abi_def>(), abi_serializer::create_yield_function( abi_serializer_max_time ));
+      fc::variant unpacked_action_data_json = json_from_file_or_string(_action_data_file_or_str);
+      std::cout << "action data variant: " << fc::json::to_pretty_string(unpacked_action_data_json) << std::endl;
+
+      bytes packed_action_data_string;
```

Comment:
I would rename to `packed_action_data` as it is not a `string`.

Reply:
Addressed: [4817492](https://github.com/AntelopeIO/leap/pull/626/commits/48174929b2ba59cce7c53c7a6291843547e7d521)